syntax = "proto3";

package chat;

// The chat service definition
service ChatService {
  // User Management
  rpc Signup (SignupRequest) returns (UserResponse);
  rpc Login (LoginRequest) returns (UserResponse);
  rpc DeleteUser (DeleteUserRequest) returns (StatusResponse);
  rpc GetAllUsers (GetAllUsersRequest) returns (UsersResponse);
  rpc UpdateViewLimit (UpdateViewLimitRequest) returns (StatusResponse);
  rpc GetUserMessageLimit (GetUserMessageLimitRequest) returns (MessageLimitResponse);
  rpc SaveSettings (SaveSettingsRequest) returns (StatusResponse);
  
  // Chat Management
  rpc StartChat (StartChatRequest) returns (ChatResponse);
  rpc GetChats (GetChatsRequest) returns (ChatsResponse);
  rpc DeleteChats (DeleteChatsRequest) returns (StatusResponse);
  rpc DeleteMessages (DeleteMessagesRequest) returns (StatusResponse);
  rpc GetMessages (GetMessagesRequest) returns (MessagesResponse);
  rpc SendChatMessage (SendMessageRequest) returns (MessageResponse);
  rpc GetUsersToDisplay (GetUsersToDisplayRequest) returns (UsersDisplayResponse);
}

// Request/Response messages
message SignupRequest {
  string username = 1;
  string nickname = 2;
  string password = 3;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message UserResponse {
  bool success = 1;
  string error_message = 2;
  User user = 3;
}

message User {
  string username = 1;
  string nickname = 2;
  int32 message_limit = 3;
}

message DeleteUserRequest {
  string user_id = 1;
}

message GetAllUsersRequest {
  string exclude_username = 1;
}

message UsersResponse {
  repeated User users = 1;
  string error_message = 2;
}

message UpdateViewLimitRequest {
  string username = 1;
  int32 new_limit = 2;
}

message GetUserMessageLimitRequest {
  string username = 1;
}

message MessageLimitResponse {
  int32 limit = 1;
  string error_message = 2;
}

message SaveSettingsRequest {
  string username = 1;
  int32 message_limit = 2;
}

message StartChatRequest {
  string current_user = 1;
  string other_user = 2;
}

message ChatResponse {
  bool success = 1;
  string error_message = 2;
  Chat chat = 3;
}

message Chat {
  string chat_id = 1;
  repeated string participants = 2;
}

message GetChatsRequest {
  string user_id = 1;
}

message ChatsResponse {
  repeated Chat chats = 1;
  string error_message = 2;
}

message DeleteChatsRequest {
  repeated string chat_ids = 1;
}

message DeleteMessagesRequest {
  string chat_id = 1;
  repeated int32 message_indices = 2;
  string current_user = 3;
}

message GetMessagesRequest {
  string chat_id = 1;
  string current_user = 2;
}

message Message {
  string sender = 1;
  string content = 2;
  string timestamp = 3;
}

message MessagesResponse {
  repeated Message messages = 1;
  string error_message = 2;
}

message SendMessageRequest {
  string chat_id = 1;
  string sender = 2;
  string content = 3;
}

message MessageResponse {
  bool success = 1;
  string error_message = 2;
  Message message = 3;
}

message GetUsersToDisplayRequest {
  string exclude_username = 1;
  string search_pattern = 2;
  int32 current_page = 3;
  int32 users_per_page = 4;
}

message UsersDisplayResponse {
  repeated User users = 1;
  int32 total_pages = 2;
  string error_message = 3;
}

message StatusResponse {
  bool success = 1;
  string error_message = 2;
}
